src/Parsing/Impl/FStringParser.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.Python.Parsing.Ast;
+
+namespace Microsoft.Python.Parsing {
+    public class FStringParser {
+        private readonly string _fString;
+        private readonly StringBuilder _buffer;
+        private readonly List<Expression> _children;
+        private int _position;
+
+        public FStringParser(string fString) {
+            _fString = fString;
+            _position = 0;
+            _buffer = new StringBuilder();
+            _children = new List<Expression>();
+        }
+
+        public FStringExpression Parse() {
+            while (HasNextChar()) {
+                if (IsDoubleBrace()) {
+                    var brace = NextChar();
+                    Read(brace);
+                    _buffer.Append(brace);
+                } else if (PeekChar() == '{') {
+                    AddBufferedSubstring();
+                    ParseInnerExpression();
+                } else {
+                    _buffer.Append(NextChar());
+                }
+            }
+            AddBufferedSubstring();
+            return new FStringExpression(_children, _fString);
+        }
+
+        private bool IsDoubleBrace() {
+            return IsDoubleChar('{') || IsDoubleChar('}');
+        }
+
+        private bool IsDoubleChar(char next) {
+            if (_position >= _fString.Length - 1) {
+                return false;
+            }
+            var doubleNextChar = _fString[_position + 1];
+            return PeekChar() == next && doubleNextChar == next;
+        }
+
+        private void ParseInnerExpression() {
+            Read('{');
+            while (HasNextChar() && PeekChar() != '}') {
+                _buffer.Append(NextChar());
+            }
+            if (!HasNextChar()) {
+                throw new Exception();
+            }
+
+            AddStringToChildren(_buffer.ToString());
+            _buffer.Clear();
+
+            Read('}');
+        }
+
+        private void Read(char nextChar) {
+            if (PeekChar() != nextChar) {
+                throw new Exception();
+            }
+
+            NextChar();
+        }
+
+        private void AddBufferedSubstring() {
+            if (_buffer.Length == 0) {
+                return;
+            }
+
+            var openQuote = "'";
+            var x = $"{openQuote}{_buffer.ToString()}{openQuote}";
+            AddStringToChildren(x);
+            _buffer.Clear();
+        }
+
+        private void AddStringToChildren(string x) {
+            var parser = Parser.CreateParser(new MemoryStream(Encoding.UTF8.GetBytes(x)), PythonLanguageVersion.V36);
```

Comment:
The language version is definitely something you'll want to accept in the constructor from the parent parser. I think you may also just want to use a `StringReader` here (since the text has already been converted into a C# string) and don't need to make an assumption about the encoding.